loop/pagemetadata.js

```diff
@@ -0,0 +1,363 @@
+// var decrypt = require('./encrypt').decrypt;
+// var ua = require('universal-analytics');
+
+// http://stackoverflow.com/questions/22603078/syntaxerror-use-of-const-in-strict-mode
+// harmony flag??
+var DISCOVER_IMAGES_MAX  = 5;
+
+/**
+ * Get all metadata from an HTML document. This includes:
+ * - URL
+ * - title
+ * - Metadata specified in <meta> tags, including OpenGraph data
+ * - Links specified in <link> tags (short, canonical, preview images, alternative)
+ * - Content that can be found in the page content that we consider useful metadata
+ * - Microformats
+ *
+ * @param {Document} document - Document to extract data from.
+ * @param {Element} [target] - Optional element to restrict microformats lookup to.
+ * @returns {Object} Object containing the various metadata, normalized to
+ *                   merge some common alternative names for metadata.
+ */
+function getData(pageBody, documentURI) {   // , target
+
+  // console.log("getData pageBody", pageBody);
+  console.log("getData documentURI", documentURI);
+
+  var window = domino.createWindow(pageBody);
+  var document = window.document;
+
+  var result = {
+    url: _validateURL(documentURI, documentURI),
+    title: document.title,
+    previews: []
+  };
+
+  // if pushState was used to change the url, most likely all meta data is
+  // invalid. This is the case with several major sites that rely on
+  // pushState. In that case, we'll only return uri and title. If document is
+  // via XHR or something, there is no view or history.
+  // if (document.defaultView) {
+  //   var docshell = document.defaultView.QueryInterface(Ci.nsIInterfaceRequestor)
+  //     .getInterface(Ci.nsIWebNavigation)
+  //     .QueryInterface(Ci.nsIDocShell);
+  //   var shentry = {};
+  //   if (docshell.getCurrentSHEntry(shentry) &&
+  //     shentry.value && shentry.value.URIWasModified) {
+  //     return result;
+  //   }
+  // }
+
+  _getMetaData(document, documentURI, result);
+  _getLinkData(document, documentURI, result);
+  _getPageData(document, documentURI, result);
+  // result.microformats = getMicroformats(document, target);
+
+  return result;
+}
+
+/**
+ * Get metadata as defined in <meta> tags.
+ * This adds properties to an existing result object.
+ *
+ * @param {Document} document - Document to extract data from.
+ * @param {Object}  result - Existing result object to add properties to.
+ */
+function _getMetaData(document, baseUrl, result) {
+  console.log("_getMetaData");
+  // Query for standardized meta data.
+  var elements = document.querySelectorAll("head > meta[property], head > meta[name]");
+  if (elements.length < 1) {
+    return;
+  }
+  console.log("_getMetaData elements.length", elements.length);
+
+  for (var element in elements) {
+    console.log("_getMetaData element", element);
+    // console.log("_getMetaData elements[element]", elements[element]);
+    // hack to make it work, needs to be resolved
+    if (element === "item") {
```

Comment:
@Standard8  issue with stackover flow error or element has no getAttribute as it returns the item function at the end of the loop.  Need to figure out how this hack can be eliminated.